Add explanatory docs on catalog database

danielballan opened the pull request:
Goals:

- Give overview of database schema
- Capture examples motivating `parameter` and `num`, added in https://github.com/bluesky/tiled/pull/584

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+  to the Adapter
+- `management` --- enum indicating whether the data is registered `"external"` data
+  or `"writable"` data managed by Tiled
+- `structre_family` --- enum of structure types (`"container"`, `"array"`, `"table"`, ...)
```
```suggestion
- `structure_family` --- enum of structure types (`"container"`, `"array"`, `"table"`, ...)
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+- `data_uri` --- location of data, given as `file://localhost/PATH`
+  (It is planned to extend to schemes other than `file`, such as `s3`, in the
+  future.)
+- `is_directory` --- boolean
```
```suggestion
- `is_directory` --- boolean: `True` for data formats, in which the directory structure is a true internal detail, managed by some other library, e.g. Zarr, TileDB, etc; otherwise -- `False` (default).
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+integer, the Adapter will be passed a list sorted by `num`.
+
+Database triggers are used to ensure self-consistency.
+
```
```suggestion
`time_created` and `time_updated` contain timestamps related to the corresponding DB entry (Node, Data Source, Asset), and not the underlying data files.
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+`data_source_asset_assocation` table is best described by the example below.
+
+- `data_source_id`, `asset_id` --- foreign keys
+- `parameter` --- the name of the parameter this Asset should be passed to
```
```suggestion
- `parameter` --- the name of the Adapter's parameter this Asset should be passed to (e.g. `"data_uri"`, `"data_uris"`, `"filepath"`, `"test"`)
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+
+id | mimetype | parameters |
+-- | -- | --
+1 | "application/x-hdf5" | {"smwr": true} | NULL
```
```suggestion
1 | "application/x-hdf5" | {"smwr": true}
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+
+id | mimetype | parameters |
+-- | -- | --
+1 | "application/x-zarr" | {} | NULL
```
```suggestion
1 | "application/x-zarr" | {}
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+-- | -- | --
+1 | "file://localhost/path/to/data.zarr" | true
+
+(Notice is_directory is `true`.)
```
```suggestion
(Notice `is_directory` is `true`.)
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+
+id | mimetype | parameters |
+-- | -- | --
+1 | "multipart/related;type=image/tiff" | {} | NULL
```
```suggestion
1 | "multipart/related;type=image/tiff" | {}
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+1 | "multipart/related;type=image/tiff" | {} | NULL
+
+The MIME type `multipart/related;type=image/tiff` is registered to an Adapter
+that expects a _sequence_ of TIFF files.
```
```suggestion
that expects a _sequence_ of TIFF files, e.g. `TiffSequenceAdapter`.
```

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+
+The MIME type `multipart/related;type=image/tiff` is registered to an Adapter
+that expects a _sequence_ of TIFF files.
+
```
This is probably nitpicking, but I've just noticed that the `__init__` method for `TiffSequenceAdapter` doesn't have `data_uris` parameter, but a class method `from_uris` does. In contrast, in `TiffAdapter` expects `data_uri` in its `__init__`. I haven't checked the rest of the adapters yet, but I feel there might be an inconsistency in whether `parameter` relates to the normal `__init__` or a specific class method.

genematx commented on docs/source/explanations/catalog.md:
```diff
@@ -0,0 +1,291 @@
+
+id | mimetype | parameters |
+-- | -- | --
+1 | "application/x-hdf5" | {} | NULL
```
```suggestion
1 | "application/x-hdf5" | {}
```

dylanmcreynolds reviewed (commented):
Should the ER Diagram that's currently in `architecture.md` be moved into here?

dylanmcreynolds reviewed (approved):
Nice! Approved, though I left a comment about moving the ER Diagram from architecture here.

danielballan commented:
Good note. I moved the diagram (replacing it with a link in its original location). I added a one-line description of each table below the diagram. It all fits on one screen. Seems good to have.

![image](https://github.com/bluesky/tiled/assets/2279598/ba7d8ce4-6613-4283-85f8-519921637f73)

danielballan commented:
(CI failure on Python 3.11 is #715, unrelated.)
